R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+#' ints2 <- abs(rnorm(length(mzs), 10))
+#' ints2[11:20] <- c(15, 30, 60, 120, 300, 200, 90, 60, 30, 23)
+#' ints3 <- abs(rnorm(length(mzs), 10))
+#' ints3[11:20] <- c(13, 20, 50, 100, 200, 100, 80, 40, 30, 20)
+#'
+#' ## Create the spectra.
+#' sp1 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.01),
+#'     intensity = ints1)
+#' sp2 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.01),
+#'     intensity = ints2)
+#' sp3 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.009),
+#'     intensity = ints3)
+#'
+#' ## Combine the spectra
+#' sp_agg <- combineSpectra(list(sp1, sp2, sp3))
+#'
+#' ## Plot the spectra before and after combining
+#' par(mfrow = c(2, 1), mar = c(4.3, 4, 1, 1))
+#' plot(mz(sp1), intensity(sp1), xlim = range(mzs[5:25]), type = "h", col = "red")
+#' points(mz(sp2), intensity(sp2), type = "h", col = "green")
+#' points(mz(sp3), intensity(sp3), type = "h", col = "blue")
+#' plot(mz(sp_agg), intensity(sp_agg), xlim = range(mzs[5:25]), type = "h",
+#'     col = "black")
+combineSpectra <- function(x, mzFun = base::mean, intensityFun = base::mean,
+                           main = ceiling(median(1:length(x))),
+                           mzd) {
+    if (length(unique(unlist(lapply(x, function(z) z@msLevel)))) != 1)
+        stop("Can only combine spectra with the same MS level")
+    mzs <- lapply(x, function(z) z@mz)
+    mzs_lens <- base::lengths(mzs)
+    mzs <- unlist(mzs, use.names = FALSE)
+    mz_order <- base::order(mzs)
+    mzs <- mzs[mz_order]
+    mz_groups <- .group_mz_values(mzs, mzd = mzd)
+    if (length(unique(mz_groups)) < length(x[[main]]@mz))
+        stop("Got less m/z groups than m/z values in the original spectrum. ",
+             "Most likely the data is not profile-mode LCMS data.")
+    ## Want to keep only those groups with a m/z from the main spectrum.
+    keep <- unlist(lapply(
+        split(
+            rep(1:length(mzs_lens), mzs_lens)[mz_order] == main,
+            mz_groups),
+        function(z) {
+            rep(any(z), length(z))
+        }), use.names = FALSE)
+    ## Keep only values for which a m/z in main is present.
+    mz_groups <- mz_groups[keep]
+    mzs <- mzs[keep]
+    ints <- unlist(base::lapply(x, function(z) z@intensity))[mz_order][keep]
+    ## Create result.
+    new_sp <- x[[main]]
+    ## Support also weighted.mean:
+    if (is.character(mzFun) && mzFun == "weighted.mean") {
+        intsp <- split(ints, mz_groups)
+        new_sp@mz <- base::mapply(split(mzs, mz_groups), intsp,
+                                  FUN = function(mz_vals, w)
+                                      stats::weighted.mean(mz_vals, w + 1,
+                                                           na.rm = TRUE),
+                                  USE.NAMES = FALSE, SIMPLIFY = TRUE)
+        new_sp@intensity <- unlist(base::lapply(intsp, intensityFun),
+                                   use.names = FALSE)
```

Comment:
Could `intensityFun` (and/or `mzFun` in the `else` statement below) return more than `numeric(1)`? (e.g. `mean`, `sum`, `median` etc. all return `numeric(1)`) If not maybe `vapply` would be  better than `unlist(lapply(...))`.

Reply:
In theory yes, the function provided by the user could return something else. But that is not desired/expected. I'll change to `vapply` as you suggest.